### DISCUSSION TITLE
Change Sender to reference of Sender in add_emit?

### DISCUSSION THREAD

**Discussion by peter-scholtens** [fltk-rs/fltk-rs#1301] [Q&A] [answered]
If I want to expand the `Message` enumeration in [RustyEd](https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/editor.rs) by something like `Log(String)` then the compiler won't allow me this as it cannot derive a `Copy` trait. The reason for this is that the size of the message is not known at compile time and [resides on the heap instead of the stack.](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html#ways-variables-and-data-interact-clone) if I understand correctly. This could be solved with `Rc` or `Arc` but I try to keep it as simple as possible by just passing the ownership of the String.

Removing the `#[derive(Copy)]` from this `Message` enumeration solves the problem, however, introducing a new one: every call of [MenuItem::add_emit](https://docs.rs/fltk/latest/fltk/menu/struct.MenuItem.html#method.add_emit) now complains _cannot move out of `*s` which is behind a shared reference_ and _move occurs because `*s` has type `fltk::app::Sender<Message>`, which does not implement the `Copy` trait_

The latter problem could be solved by replacing all the 11 occurrences of `*s` with `s.clone()` in this editor example, but this looks a little silly and too verbose to me. Cloning has the association of creating a lot of work, but in this case it isn't because it's only a pointer? (I'm not sure though)

Therefore, my question, wouldn't it be more convenient to change the function signature of `add_emit` to include the reference to a sender like `sender: &Sender<T>` instead of consuming the sender?

**Comment by MoAlyousef** [comment, marked as answer]
The problem is that the sender is moved into the callback since callbacks require `FnMut() + 'static`, so if the signature is changed to require a reference, it will have to be a static reference to the sender and it's quite limiting.
Now senders are actually empty structs (just containing phantom data for the generic), so cloning senders is really cheap.
